lib/semantics/type.h

```diff
@@ -237,43 +253,36 @@ class DeclTypeSpec {
   };
 
   // intrinsic-type-spec or TYPE(intrinsic-type-spec), not character
-  DeclTypeSpec(const NumericTypeSpec &);
-  DeclTypeSpec(const LogicalTypeSpec &);
+  DeclTypeSpec(NumericTypeSpec &&);
+  DeclTypeSpec(LogicalTypeSpec &&);
   // character
-  DeclTypeSpec(CharacterTypeSpec &);
+  DeclTypeSpec(const CharacterTypeSpec &);
+  DeclTypeSpec(CharacterTypeSpec &&);
   // TYPE(derived-type-spec) or CLASS(derived-type-spec)
-  DeclTypeSpec(Category, DerivedTypeSpec &);
+  DeclTypeSpec(Category, const DerivedTypeSpec &);
+  DeclTypeSpec(Category, DerivedTypeSpec &&);
   // TYPE(*) or CLASS(*)
   DeclTypeSpec(Category);
-  DeclTypeSpec() = delete;
 
   bool operator==(const DeclTypeSpec &) const;
   bool operator!=(const DeclTypeSpec &that) const { return !operator==(that); }
 
   Category category() const { return category_; }
   bool IsNumeric(TypeCategory) const;
+  IntrinsicTypeSpec *AsIntrinsic();
   const IntrinsicTypeSpec *AsIntrinsic() const;
   const DerivedTypeSpec *AsDerived() const;
   const NumericTypeSpec &numericTypeSpec() const;
   const LogicalTypeSpec &logicalTypeSpec() const;
   const CharacterTypeSpec &characterTypeSpec() const;
   const DerivedTypeSpec &derivedTypeSpec() const;
   DerivedTypeSpec &derivedTypeSpec();
```

Comment:
For a future change, I wonder if it would be a better interface now if it were similar to Symbol, with member functions like:
```
template<typename T> bool has() const;
template<typename T> const T &get() const;
template<typename T> const T *get_if() const;
```
We would have to add some classes to the variant but then could eliminate the category.

Reply:
Right, good idea, we'd just have to use two distinct empty structs for TYPE(*) and CLASS(*) alternatives.

Reply:
Markdown removed the asterisks from the parentheses in my comment above, sorry.  I meant `TYPE(*)` and `CLASS(*)`.